crates/aptos-jwk-consensus/src/network.rs

```diff
@@ -0,0 +1,184 @@
+            .map_err(RpcError::ApplicationError);
+        if let Some(tx) = self.inner.take() {
+            let _ = tx.send(rpc_response);
+        }
+    }
+}
+
+#[cfg(test)]
+pub struct DummyRpcResponseSender {
+    pub rpc_response_collector: Arc<RwLock<Vec<anyhow::Result<JWKConsensusMsg>>>>,
+}
+
+#[cfg(test)]
+impl DummyRpcResponseSender {
+    pub fn new(rpc_response_collector: Arc<RwLock<Vec<anyhow::Result<JWKConsensusMsg>>>>) -> Self {
+        Self {
+            rpc_response_collector,
+        }
+    }
+}
+
+#[cfg(test)]
+impl RpcResponseSender for DummyRpcResponseSender {
+    fn send(&mut self, response: anyhow::Result<JWKConsensusMsg>) {
+        self.rpc_response_collector.write().push(response);
+    }
+}
+
+pub struct NetworkReceivers {
+    pub rpc_rx: aptos_channel::Receiver<AccountAddress, (AccountAddress, IncomingRpcRequest)>,
+}
+
+pub struct NetworkTask {
+    all_events: Box<dyn Stream<Item = Event<JWKConsensusMsg>> + Send + Unpin>,
+    rpc_tx: aptos_channel::Sender<AccountAddress, (AccountAddress, IncomingRpcRequest)>,
+}
+
+impl NetworkTask {
+    /// Establishes the initial connections with the peers and returns the receivers.
+    pub fn new(
+        network_service_events: NetworkServiceEvents<JWKConsensusMsg>,
+        self_receiver: aptos_channels::Receiver<Event<JWKConsensusMsg>>,
+    ) -> (NetworkTask, NetworkReceivers) {
+        let (rpc_tx, rpc_rx) = aptos_channel::new(QueueStyle::FIFO, 10, None);
+
+        let network_and_events = network_service_events.into_network_and_events();
+        if (network_and_events.values().len() != 1)
+            || !network_and_events.contains_key(&NetworkId::Validator)
+        {
+            panic!("The network has not been setup correctly for JWK consensus!");
+        }
+
+        // Collect all the network events into a single stream
+        let network_events: Vec<_> = network_and_events.into_values().collect();
+        let network_events = select_all(network_events).fuse();
+        let all_events = Box::new(select(network_events, self_receiver));
+
+        (NetworkTask { rpc_tx, all_events }, NetworkReceivers {
+            rpc_rx,
+        })
+    }
+
+    pub async fn start(mut self) {
+        while let Some(message) = self.all_events.next().await {
+            match message {
+                Event::RpcRequest(peer_id, msg, protocol, response_sender) => {
+                    let req = IncomingRpcRequest {
+                        msg,
+                        sender: peer_id,
+                        response_sender: Box::new(RealRpcResponseSender {
+                            inner: Some(response_sender),
+                            protocol,
+                        }),
+                    };
+
+                    if let Err(e) = self.rpc_tx.push(peer_id, (peer_id, req)) {
+                        warn!(error = ?e, "aptos channel closed");
+                    };
+                },
+                _ => {
+                    // Ignore
```

Comment:
@grao1991 why sampled log instead of ignoring?

Reply:
I guess you don't expect receiving other things here, if that happens it's an indication of bug and you want to know it here.